src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Formatters/CollectionFormatter.cs

```diff
@@ -414,18 +414,22 @@ public void Serialize(ref MessagePackWriter writer, TCollection? value, MessageP
         // abstraction for serialize
         protected virtual int? GetCount(TCollection sequence)
         {
+#if NET6_0_OR_GREATER
+            if (Enumerable.TryGetNonEnumeratedCount(sequence, out var count))
+            {
+                return count;
+            }
+#else
             var collection = sequence as ICollection<TElement>;
             if (collection != null)
             {
                 return collection.Count;
             }
-            else
+#endif
+            var c2 = sequence as IReadOnlyCollection<TElement>;
```

Comment:
Surely `TryGetNonEnumeratedCount` has already tried this downcast, hasn't it? Shouldn't this attempt also be inside the `#else` block?

Reply:
[According to dotnet/runtime](https://github.com/dotnet/runtime/blob/7a60900cd1be3361773237d0d0d0293146db5547/src/libraries/System.Linq/src/System/Linq/Count.cs#L95)

`TryGetNonEnumeratedCount<TSource>` does these downcasts.

- `ICollection<TSource>`
- `IIListProvider<TSource>`
- `ICollection`

There seems no downcast to `IReadOnlyCollection<TSource>`.
So I leave checks for `IReadOnlyCollection<TSource>` in .NET 6 or later.

Reply:
That's surprising. Thanks for being so meticulous.